Constructors for compressed and uncompressed ECDSA keys

dr-orlovsky opened the pull request.

RCasatta commented on src/util/key.rs:
```diff
@@ -78,6 +78,23 @@ pub struct PublicKey {
 
 impl PublicKey {
+    /// Constructs compressed ECDSA PublicKey from the provided generic Secp256k1 public key
+    pub fn compressed(pk: secp256k1::PublicKey) -> PublicKey {
```
I would prefer just `new` for a compressed key (same for private)

dr-orlovsky replied:
Will agree. May be also `legacy` for `uncompressed`?

RCasatta replied:
For uncompressed I personally haven't a preference both uncompressed or legacy are ok

apoelstra replied:
I'd prefer `new` and `new_uncompressed`

dr-orlovsky commented:
@apoelstra done

sanket1729 reviewed (commented):
@dr-orlovsky, I think you probably forgot to push the changes :stuck_out_tongue:

dr-orlovsky commented:
@sanket1729 yeah, also needed rebase baeause of #589 merge

sgeisler commented on src/util/ecdsa.rs:
```diff
@@ -37,6 +37,23 @@ pub struct PublicKey {
 }
 
 impl PublicKey {
+    /// Constructs compressed ECDSA PublicKey from the provided generic Secp256k1 public key
```
Very minor nit: if it's not code, i.e. a type, I wouldn't write public key in camel case. If it's code/a type it should be formatted as such.

sgeisler commented on src/util/ecdsa.rs:
```diff
@@ -37,6 +37,23 @@ pub struct PublicKey {
+    pub fn new(pk: secp256k1::PublicKey) -> PublicKey {
+        PublicKey {
+            compressed: true,
+            key: pk,
```
Minor nit, `pk` could be renamed to key which would allow to write `key,` instead of `key: pk,`.

sgeisler commented on src/util/ecdsa.rs:
```diff
@@ -151,6 +168,26 @@ pub struct PrivateKey {
+    pub fn new(pk: secp256k1::SecretKey, network: Network) -> PrivateKey {
+        PrivateKey {
+            compressed: true,
+            network: network,
```
Why not just `network,`?

sanket1729 replied:
I don't know why but I prefer `network: network` instead of `network`, looks aesthetically cleaner to me when you have other fields that use the `a: b` syntax.

sgeisler replied:
Ok, I probably just got conditioned by my linter :smile:

dr-orlovsky replied:
I was always simplifying `A: A` to `A`, but with the ancient rustc 1.29 I am not sure whether it knows about this stuff – and I am tired of CI failing just because you use rust syntactic sugar.

sgeisler reviewed (approved):
utACK 6e9d5494181268b1d16ebfc5c8aa337131a6b409

I think this can be merged if you have a good use case for it as-is. I wonder though if we shouldn't take a look at where these new functions could already be used in our library.

dr-orlovsky commented:
@sgeisler I checked the library, updated all places (including tests and docs) where the new API can be used. Next, as a separate commit 187eae8 I added `PrivateKey::from_slice` convenience method, since there were multiple use of `SecretKey::from_slice` with a boilerplate code around it. Additionally, I removed unnecessary `map_err`, which simplified readability of BIP32-related methods.

sgeisler reviewed (approved):
utACK 187eae8a13d958e21a879bf4ad95c95416b43c0a

sanket1729 reviewed (approved):
ACk 187eae8a13d958e21a879bf4ad95c95416b43c0a

apoelstra reviewed (approved):
ack 187eae8a13d958e21a879bf4ad95c95416b43c0a
